src/main/java/com/rabbitmq/jms/client/DelayedMessageService.java

```diff
@@ -0,0 +1,72 @@
+package com.rabbitmq.jms.client;
+
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.jms.admin.RMQDestination;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Semaphore;
+
+class DelayedMessageService {
+    static final String X_DELAYED_JMS_EXCHANGE = "x-delayed-jms-message";
+    static final String X_DELAY_HEADER = "x-delay";
+    static final String X_DELAYED_JMS_EXCHANGE_HEADER = "delayed-exchange";
+
+    /** Cache of exchanges which have been bound to the delayed exchange */
+    private Map<String, Boolean> delayedDestinations;
+    private volatile boolean delayedExchangeDeclared;
+    private Semaphore declaring = new Semaphore(1);
+
+    public DelayedMessageService() {
+        this.delayedDestinations = new ConcurrentHashMap<>();
+    }
+    public void close() {
+        delayedDestinations.clear();
+    }
+
+    public String delayMessage(Channel channel, RMQDestination destination, Map<String, Object> messageHeaders, long deliveryDelayMs) {
+        if (deliveryDelayMs <= 0L) return destination.getAmqpExchangeName();
+
+        declareDelayedExchange(channel);
+        delayedDestinations.computeIfAbsent(destination.getAmqpExchangeName(), destination1 -> {
+            bindDestinationToDelayedExchange(channel, destination);
+            return true;
+        });
+        messageHeaders.put(X_DELAY_HEADER, deliveryDelayMs);
+        messageHeaders.put(X_DELAYED_JMS_EXCHANGE_HEADER, destination.getAmqpExchangeName());
+        return X_DELAYED_JMS_EXCHANGE;
+    }
+    private void declareDelayedExchange(Channel channel) {
```

Comment:
You can use an `AtomicBoolean` initialized with `false` (not declared) and surround the declaration with `AtomicBoolean#compareAndSet(false, true)` to make it idempotent. You should not need the semaphore then.

Reply:
I use a semaphore so that the second entrant call has to wait until the exchange is declared by the first entrant. It is not only to protect the critical section.